Madness/String.swift

```diff
@@ -0,0 +1,78 @@
+//
+//  String.swift
+//  Madness
+//
+//  Created by Josh Vera on 10/19/15.
+//  Copyright © 2015 Rob Rix. All rights reserved.
+//
+
+import Foundation
+import Prelude
+
+public typealias CharacterParser = Parser<String.CharacterView, Character>.Function
+public typealias CharacterArrayParser = Parser<String.CharacterView, [Character]>.Function
+public typealias StringParser = Parser<String.CharacterView, String>.Function
+public typealias DoubleParser = Parser<String.CharacterView, Double>.Function
+
+public typealias IntParser = Parser<String.CharacterView, Int>.Function
+
+func maybePrepend<T>(value: T?) -> [T] -> [T] {
+	return { value != nil ? [value!] + $0 : $0 }
+}
+
+func concat<T>(value: [T]) -> [T] -> [T] {
+	return { value + $0 }
+}
+
+func concat2<T>(value: [T])(value2: [T]) -> [T] -> [T] {
+	return { value + value2 + $0 }
+}
+
+let someDigits: CharacterArrayParser = some(digit)
+
+public let int: CharacterArrayParser = {
+	let minus: Parser<String.CharacterView, Character?>.Function = char("-")|?
+	
+	return maybePrepend <^> minus <*> someDigits
+}()
+
+
+let decimal: CharacterArrayParser = prepend <^> %"." <*> someDigits
+
+let exp: StringParser = %"e" <|> %"e+" <|> %"e-" <|> %"E" <|> %"E+" <|> %"E-"
+
+let exponent: CharacterArrayParser = { s in { s.characters + $0 } } <^> exp <*> someDigits
+
+public let number: DoubleParser = { characters in Double(String(characters))! } <^>
+	(int
+	<|> (concat <^> int <*> decimal)
+	<|> (concat <^> int <*> exponent)
+	<|> (concat2 <^> int <*> decimal <*> exponent))
```

Comment:
Would this last expression be equivalent to the composition of `concat` on itself?

```swift
(concat >>> concat) <^> int <*> decimal <*> exponent
```